### PR TITLE
APIMF-2544: added $defs parsing and emission for Draft 2019-09. Also setup up everything to start parsing the draft.

### PR DESCRIPTION
**Pull request by tomsfernandez**

### CODE REVIEW

File: amf-webapi/shared/src/main/scala/amf/plugins/document/webapi/parser/spec/jsonschema/AstFinder.scala [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+import amf.plugins.document.webapi.contexts.WebApiContext
+import amf.plugins.document.webapi.contexts.parser.oas.JsonSchemaWebApiContext
+import amf.plugins.document.webapi.contexts.parser.raml.Raml08WebApiContext
+import amf.plugins.document.webapi.parser.spec.common.YMapEntryLike
+import amf.plugins.document.webapi.parser.spec.toOasDeclarations
+import amf.validations.ParserSideValidations.UnableToParseJsonSchema
+import org.yaml.model.{YDocument, YMap, YNode}
+import org.yaml.parser.{YParser, YamlParser}
+
+object AstFinder {
```
Comment (nschejtman) [inline amf-webapi/shared/src/main/scala/amf/plugins/document/webapi/parser/spec/jsonschema/AstFinder.scala:12]: Seems this object has more responsibilities than finding nodes in an AST

File: amf-webapi/shared/src/main/scala/amf/plugins/document/webapi/parser/spec/declaration/SchemaVersion.scala [hunk 2]
```diff
@@ -4,33 +4,44 @@ import amf.client.render.{JSONSchemaVersions => ClientJSONSchemaVersions, JSONSc
...
+abstract class SchemaVersion(val name: String)
+
+object SchemaVersion {
+  def fromClientOptions(schema: ClientJSONSchemaVersion): SchemaVersion = schema match {
+    case ClientJSONSchemaVersions.DRAFT_07 => JSONSchemaDraft7SchemaVersion
+    case _                                 => JSONSchemaDraft4SchemaVersion
+  }
+}
+
+abstract class RAMLSchemaVersion(override val name: String) extends SchemaVersion(name)
```
Comment (nschejtman) [inline amf-webapi/shared/src/main/scala/amf/plugins/document/webapi/parser/spec/declaration/SchemaVersion.scala:23]: [Out of scope of this story] Why do we version RAML & OAS Schemas? Not clear why de we need to synthesize a RAML/OAS schema version? After this doing this story, did you see any use for these?

File: amf-webapi/shared/src/main/scala/amf/plugins/domain/shapes/resolution/stages/merge/JsonMergePatchAlgorithm.scala [hunk 3]
```diff
@@ -45,6 +45,7 @@ case class JsonMergePatch(isNull: AmfElement => Boolean, keyCriteria: KeyCriteri
     patch.fields
       .fields()
       .filter(entry => !ignoredFields.contains(entry.field))
+
```
Comment (nschejtman) [inline amf-webapi/shared/src/main/scala/amf/plugins/domain/shapes/resolution/stages/merge/JsonMergePatchAlgorithm.scala]: Watch out for empty lines

Comment (tomsfernandez) [reply amf-webapi/shared/src/main/scala/amf/plugins/domain/shapes/resolution/stages/merge/JsonMergePatchAlgorithm.scala]: Will fix it.

File: amf-webapi/shared/src/main/scala/amf/plugins/document/webapi/parser/spec/declaration/SchemaVersion.scala (hunk 2 above)
Comment (tomsfernandez) [reply amf-webapi/shared/src/main/scala/amf/plugins/document/webapi/parser/spec/declaration/SchemaVersion.scala:23]: OAS versions are primarily used due to a lack of correct design in the OasTypeParser. OAS Schema Objects have some differences with Json Schema ones, primarily in validation. That is why we version them.
RAML version was used as a config setting to parse Raml Types in Async.

Ideally, we would only need them to know which TypeParser to build (this mostly applies to OAS and JsonSchema). They should be ValueObjects with no other responsibility than identifying the parser I wish to build and parse with.

File: amf-webapi/shared/src/main/scala/amf/plugins/document/webapi/parser/spec/jsonschema/AstFinder.scala (hunk 1 above)
Comment (tomsfernandez) [reply amf-webapi/shared/src/main/scala/amf/plugins/document/webapi/parser/spec/jsonschema/AstFinder.scala:12]: Added a commit that will solve this issue

**Review by nschejtman — APPROVED** [review]

**Pull request merged**